waku/v2/api/filter.go

```diff
@@ -0,0 +1,225 @@
+package api
+
+import (
+	"context"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/waku-org/go-waku/waku/v2/protocol"
+	"github.com/waku-org/go-waku/waku/v2/protocol/filter"
+	"github.com/waku-org/go-waku/waku/v2/protocol/subscription"
+	"go.uber.org/zap"
+	"golang.org/x/exp/maps"
+)
+
+type FilterConfig struct {
+	MaxPeers int
+	Peers    []peer.ID
+}
+
+type Sub struct {
+	ContentFilter protocol.ContentFilter
+	DataCh        chan *protocol.Envelope
+	Config        FilterConfig
+	subs          subscription.SubscriptionSet
+	wf            *filter.WakuFilterLightNode
+	ctx           context.Context
+	cancel        context.CancelFunc
+	log           *zap.Logger
+}
+
+// Subscribe
+func Subscribe(ctx context.Context, wf *filter.WakuFilterLightNode, contentFilter protocol.ContentFilter, config FilterConfig) (*Sub, error) {
+	sub := new(Sub)
+	sub.wf = wf
+	sub.ctx, sub.cancel = context.WithCancel(ctx)
+	sub.subs = make(subscription.SubscriptionSet)
+	sub.DataCh = make(chan *protocol.Envelope)
+	sub.ContentFilter = contentFilter
+	sub.Config = config
+	sub.log = func() *zap.Logger {
+		log, _ := zap.NewDevelopment()
+		return log
+	}().Named("filterv2-api")
+
+	subs, err := sub.subscribe(contentFilter, sub.Config.MaxPeers)
+
+	if err == nil {
+		sub.multiplex(subs)
+		sub.log.Info("go sub.healthCheckLoop()")
+		go sub.healthCheckLoop()
+		return sub, nil
+	} else {
+		return nil, err
+	}
+}
+
+func (apiSub *Sub) Unsubscribe() {
+	apiSub.cancel()
+
+}
+
+func (apiSub *Sub) healthCheckLoop() {
+	// Health checks
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-apiSub.ctx.Done():
+			apiSub.log.Info("healthCheckLoop: Done()")
+			apiSub.cleanup()
+			return
+		case <-ticker.C:
+			apiSub.log.Info("healthCheckLoop: checkAliveness()")
+			topicCounts := apiSub.getTopicCounts()
+			apiSub.resubscribe(topicCounts)
+		}
+	}
+
+}
+
+func (apiSub *Sub) cleanup() {
+	apiSub.log.Info("ENTER cleanup()")
+	defer func() {
+		apiSub.log.Info("EXIT cleanup()")
+	}()
+
+	for _, s := range apiSub.subs {
+		apiSub.wf.UnsubscribeWithSubscription(apiSub.ctx, s)
+	}
+	close(apiSub.DataCh)
+
+}
+
+// Returns active sub counts for each pubsub topic
+func (apiSub *Sub) getTopicCounts() map[string]int {
+	apiSub.log.Info("ENTER getTopicCounts()")
+	defer func() {
+		apiSub.log.Info("EXIT getTopicCounts()")
+	}()
+
+	// Buffered chan for sub aliveness results
+	type CheckResult struct {
+		sub   *subscription.SubscriptionDetails
+		alive bool
+	}
+	checkResults := make(chan CheckResult, len(apiSub.subs))
+
+	// Run pings asynchronously
+	for _, s := range apiSub.subs {
```

Comment:
Don't we need to lock this map? What if a parallel healthCheck and subscribe/unsubscribe gets invoked. 
Since we are iterating, it would throw a concurrent access error.

Reply:
We don't, unless i'm missing something. If you look at the code, each iteration through `apiSub.subs` occurs in the same goroutine - the one that runs `healthCheckLoop()`. There are no race conditions.

Reply:
Ah, right...for some reason i was thinking all subs are stored in one set. 
Each subscribe is having its own object...which is why it is fine.
No locking needed.